app/src/main/java/com/example/habitsmasher/UserValidator.java

```diff
@@ -0,0 +1,155 @@
+    private static final int LOWER_CHARACTER_LIMIT = 6;
+
+    private EditText _usernameInput;
+    private EditText _emailInput;
+    private EditText _passwordInput;
+
+    String _username;
+    String _email;
+    String _password;
+
+    // for testing
+    public UserValidator() {}
+
+    // for sign up
+    public UserValidator(EditText usernameInput,
+                         EditText emailInput,
+                         EditText passwordInput) {
+        _usernameInput = usernameInput;
+        _emailInput = emailInput;
+        _passwordInput = passwordInput;
+
+        // save text contents from EditText fields
+        _email = _emailInput.getText().toString().trim();
+        _password = _passwordInput.getText().toString().trim();
+        _username = _usernameInput.getText().toString().trim();
+    }
+
+    // for log in
+    public UserValidator(EditText emailInput,
+                         EditText passwordInput) {
+        _usernameInput = null;
+        _emailInput = emailInput;
+        _passwordInput = passwordInput;
+
+        // save text contents from EditText fields
+        _email = _emailInput.getText().toString().trim();
+        _password = _passwordInput.getText().toString().trim();
+        _username = "";  // since there is no username for logging in, empty
+    }
+
+    public String getValidUsernameForSignUp() {
+        if (isNewUserValid()) {
+            return _username;
+        }
+        return "";
+    }
+    public String getValidEmailForSignUp() {
+        if (isNewUserValid()) {
+            return _email;
+        }
+        return "";
+    }
+    public String getValidPasswordForSignUp() {
+        if (isNewUserValid()) {
+            return _password;
+        }
+        return "";
+    }
+
+    public String getValidEmailForLogin() {
+        if (isLoginValid()) {
+            return _email;
+        }
+        return "";
+    }
+    public String getValidPasswordForLogin() {
+        if (isLoginValid()) {
+            return _password;
+        }
+        return "";
+    }
+
+    private boolean isNewUserValid() {
+        return isNewUserValid(_username, _email, _password);
+    }
+
+    private boolean isLoginValid() {
+        return isLoginValid(_email, _password);
+    }
+
+    public boolean isNewUserValid(String username, String email, String password) {
+        return !(username.isEmpty() | username.length() > UPPER_CHARACTER_LIMIT | !isLoginValid(email, password));
+    }
+
+    public boolean isLoginValid(String email, String password) {
+        return !(email.isEmpty() | password.isEmpty() |
+                !PatternsCompat.EMAIL_ADDRESS.matcher(email).matches() |
+                password.length() < LOWER_CHARACTER_LIMIT |
+                password.length() > UPPER_CHARACTER_LIMIT);
+    }
+
+
+    public void showSignUpErrors() {
+        if (_username.isEmpty()) {
+            _usernameInput.setError(USERNAME_IS_REQUIRED_MESSAGE);
+            _usernameInput.requestFocus();
```

Comment:
I can sorta guess what `setError` does, but what does `requestFocus` do? I remember seeing a bunch of StackOverflow people talking about focus, but I never really understood it.

Reply:
Good question, `requestFocus` sets the cursor to the specified field. So in this case, we want the user to **fix** the error, so this conveniently brings the cursor to the incorrect field.